platooning_strategic_IHP/src/platoon_strategic_ihp.cpp

```diff
@@ -1945,35 +2806,154 @@ namespace platoon_strategic_ihp
-            ROS_DEBUG_STREAM("end_dist: " << end_dist);
-            double dist_diff = end_dist - current_progress;
-            ROS_DEBUG_STREAM("dist_diff: " << dist_diff);
-            if(end_dist < current_progress){
-                break;
-            }
+        // lane change maneuver 
+        if (pm_.safeToLaneChange)
+        {   
+            // for testing purpose only, check lane change status
+            double start_crosstrack = 0.5*config_.laneWidth; // Assume vehicle start at left lane when testing.
+            double crosstrackDiff = current_crosstrack_ - start_crosstrack; 
+            bool isLaneChangeFinished = crosstrackDiff >= config_.laneWidth*0.85; // Use 85% of lane width to account for noise.
+             
+            /**  
+             * Note: The function "find_target_lanelet_id" was used to test the IHP platooning logic and is only a pre-written scenario. 
+             *       We re-use the existing lane change maneuver for route-following, and followed the data flow just to compile the code. 
+             * 
+             * TODO: The IHP2 platooning should provide necessary data in a maneuver plan for the arbitrary lane change module.
+             *       This means this entire block need to be updated accordingly. 
+             */ 
+            
+            // lane change not finished, use lane change plan
+            if(!isLaneChangeFinished)  
+            {
+                // send out lane change plan
+                while(current_progress < total_maneuver_length)
+                {   
+                    ROS_DEBUG_STREAM("Lane Change Maneuver for Cut-in join ! ");
+                    ROS_DEBUG_STREAM("Lanlet: " << shortest_path[last_lanelet_index].id());
+                    ROS_DEBUG_STREAM("current_progress: "<< current_progress);
+                    ROS_DEBUG_STREAM("speed_progress: " << speed_progress);
+                    ROS_DEBUG_STREAM("target_speed: " << target_speed);
+                    ROS_DEBUG_STREAM("time_progress: " << time_progress.toSec());
+
+                    // ----------------------- UCLA: consider change according to maneuver plan requirement --------------------
+                    auto p = shortest_path[last_lanelet_index].centerline2d().back(); // change to lane change path
+                    // set to next lane destination, consider sending ecef instead of dtd 
+                    double end_dist = wm_->routeTrackPos(shortest_path[last_lanelet_index].centerline2d().back()).downtrack;
+                    end_dist = std::min(end_dist, total_maneuver_length);
+                    ROS_DEBUG_STREAM("end_dist: " << end_dist);
+                    // consider calculate dtd_diff and ctd_diff
+                    double dist_diff = end_dist - current_progress;
+                    ROS_DEBUG_STREAM("dist_diff: " << dist_diff);
+                    // UCLA: start and end lane IDs
+                    int current_lanelet_id = shortest_path[last_lanelet_index].id();
```

Comment:
your current lanelet id should be determined using the pose, not from list of shortest path

Reply:
Updated. We can use current_lanelets list (which was determined based on vehicle pose) to read current lanelet_ID.